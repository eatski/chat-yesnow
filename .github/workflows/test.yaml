name: Integration Tests

on: [push]

jobs:
  test:
    runs-on: ubuntu-latest
    env:
      SUPABASE_POSTGRES_URL_NON_POOLING: postgresql://postgres:@localhost:5432/postgres
      SUPABASE_POSTGRES_PRISMA_URL: postgresql://postgres:@localhost:5432/postgres?pgbouncer=true
      CI: 1
    services:
      postgres:
        image: postgres:13-alpine
        # refer: https://hub.docker.com/_/postgres
        env:
          POSTGRES_PASSWORD: postgres
          POSTGRES_USER: postgres
          POSTGRES_DB: postgres
        ports:
          - 5432:5432

    steps:
      - name: Check out code
        uses: actions/checkout@v2

      - name: Setup Node.js
        uses: actions/setup-node@v2
        with:
          node-version: 16

      - name: Install dependencies
        run: npm install

      - name: Run Prisma Migrate
        run: npx prisma migrate deploy
      
      - name: Prepare Test Data
        run: npm run test:prepare
      
      - name: Run Tests
        run: npm run test
      
        
