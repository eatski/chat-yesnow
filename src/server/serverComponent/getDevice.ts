import { Device, getDevice } from "@/common/util/device";
import { headers } from "next/headers";

export const getDeviceServer = (): Device => {
	// User Agent の解析
	const ua = headers().get("user-agent");
	return getDevice(ua || undefined);
};
