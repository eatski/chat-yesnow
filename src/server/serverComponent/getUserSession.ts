import { authConfig } from "@/pages/api/auth/[...nextauth]";
import { getServerSession } from "next-auth";

export type UserSession = {
	email: string;
	userId: string;
};

export const getUserSession = async (): Promise<UserSession | null> => {
	const session = await getServerSession(authConfig);
	console.log("session", session);
	if (!session || !session.user?.email || !session.custom?.userId) {
		return null;
	}
	return {
		email: session.user.email,
		userId: session.custom.userId,
	};
};
