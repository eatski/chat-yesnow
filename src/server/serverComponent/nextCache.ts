import { unstable_cache } from "next/cache";

type StringifyObject<T> = T extends Date
	? string
	: T extends (infer U)[]
		? StringifyObject<U>[]
		: T extends object
			? { [K in keyof T]: StringifyObject<T[K]> }
			: T;

type CacheFnParameters = Parameters<typeof unstable_cache>;
type Tail<T extends any[]> = T extends [any, ...infer Rest] ? Rest : never;

export const nextCache = <T>(
	fn: () => Promise<T>,
	...tailParameters: Tail<CacheFnParameters>
): (() => Promise<StringifyObject<T>>) => {
	// @ts-ignore
	return unstable_cache(fn, ...tailParameters);
};
