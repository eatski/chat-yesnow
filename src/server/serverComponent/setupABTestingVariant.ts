import {
	AB_TESTING_COOKIE_NAME,
	AB_TESTING_VARIANTS,
	getAorBRandom,
} from "@/common/abtesting";
import { get } from "@vercel/edge-config";
import { cookies } from "next/headers";

export const setupABTestValue = async () => {
	const abTestRate = await get("abTestRate").catch(() => null);
	if (
		!abTestRate ||
		!(typeof abTestRate === "number") ||
		Number.isNaN(abTestRate) ||
		abTestRate < 0 ||
		abTestRate > 1
	) {
		cookies().delete(AB_TESTING_COOKIE_NAME);
		return AB_TESTING_VARIANTS.ONLY_SONNET;
	}
	const random = getAorBRandom(abTestRate);
	cookies().set(AB_TESTING_COOKIE_NAME, random, {
		httpOnly: false,
		secure: true,
		sameSite: "none",
	});
	return random;
};
