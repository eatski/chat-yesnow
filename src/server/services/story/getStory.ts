import { revalidateTime } from "../../../common/revalidate";
import { prisma } from "../../../libs/prisma";
import type { Story } from "../../../server/model/story";
import { nextCache } from "../../../server/serverComponent/nextCache";
import { hydrateStory } from "./functions";

type FilterForFindFirst =
	| {
			type: "onlyPublic";
	  }
	| {
			type: "publicOrWithAuthor" | "withAuthorId";
			authorId: string;
	  };

const cacheKeyByFilter = (filter: FilterForFindFirst): string[] => {
	const simpleTypedFilter: {
		type: FilterForFindFirst["type"];
		authorId?: string;
	} = filter;
	return [simpleTypedFilter.type, simpleTypedFilter.authorId].filter(
		(v) => v != null,
	);
};

export const getStory = ({
	storyId,
	filter,
}: {
	storyId: string;
	filter: FilterForFindFirst;
}): Promise<Story | null> => {
	return nextCache(
		async () => {
			const story = await prisma.story.findFirst({
				where: {
					id: storyId,
				},
				include: {
					author: {
						select: {
							id: true,
							name: true,
						},
					},
				},
			});

			if (story == null) return null;

			if (filter.type === "withAuthorId" && story.authorId !== filter.authorId)
				return null;
			if (filter.type === "onlyPublic" && !story.published) return null;
			if (
				filter.type === "publicOrWithAuthor" &&
				!story.published &&
				story.authorId !== filter.authorId
			)
				return null;

			return hydrateStory(story);
		},
		["_getStory", storyId, ...cacheKeyByFilter(filter)],
		{
			revalidate: revalidateTime.short,
			tags: [`/stories/${storyId}`],
		},
	)();
};
