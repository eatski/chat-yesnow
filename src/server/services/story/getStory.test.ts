import {
	afterEach,
	beforeAll,
	beforeEach,
	describe,
	expect,
	it,
	test,
	vi,
	vitest,
} from "vitest";

import { resolveFixturePath } from "../../../../fixtures";
import { prepareStoryFromYaml } from "../../../test/prepareStory";
import * as nextCache from "../../serverComponent/nextCache";
import { getStory } from "./getStory";

describe("getStory", () => {
	beforeEach(() => {
		vi.spyOn(nextCache, "nextCache").mockImplementation((fn) => fn);
	});
	describe("onlyPublic", () => {
		test("should return the story if it is public", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: true,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);

			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "onlyPublic" },
			});
			expect(story).toMatchSnapshot();
		});
		test("should return null if the story is not public", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: false,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);
			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "onlyPublic" },
			});
			expect(story).toBe(null);
		});
	});
	describe("publicOrWithAuthor", () => {
		test("should return the story if it is public", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: true,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);

			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "publicOrWithAuthor", authorId },
			});
			expect(story).toMatchSnapshot();
		});
		test("should return the story if it is authored by the author", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: false,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);

			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "publicOrWithAuthor", authorId },
			});
			expect(story).toMatchSnapshot();
		});
		test("should return null if the story is not public and not authored by the author", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: false,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);

			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "publicOrWithAuthor", authorId: "otherAuthorId" },
			});
			expect(story).toBe(null);
		});
	});
	describe("withAuthorId", () => {
		test("should return the story if it is authored by the author", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: false,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);

			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "withAuthorId", authorId },
			});
			expect(story).toMatchSnapshot();
		});
		test("should return null if the story is not authored by the author", async (test) => {
			const storyId = test.task.name + "-storyId";
			const authorId = test.task.name + "-authorId";
			const cleanup = await prepareStoryFromYaml(
				resolveFixturePath("test.yaml"),
				{
					authorId,
					storyId,
					published: false,
					date: new Date("2024-08-28T13:54:41.065Z"),
				},
			);

			test.onTestFinished(() => {
				cleanup();
			});

			const story = await getStory({
				storyId,
				filter: { type: "withAuthorId", authorId: "otherAuthorId" },
			});

			expect(story).toBe(null);
		});
	});
});
