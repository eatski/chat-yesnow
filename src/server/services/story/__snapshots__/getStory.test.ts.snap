// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getStory > onlyPublic > should return the story if it is public 1`] = `
{
  "author": {
    "id": "should return the story if it is public-authorId",
    "name": null,
  },
  "id": "should return the story if it is public-storyId",
  "published": true,
  "publishedAt": 1724853281065,
  "questionExamples": [
    {
      "answer": "True",
      "customMessage": "いい質問ですね！山田さんは人を殺しています。",
      "question": "山田さんは人を殺害しましたか？",
    },
    {
      "answer": "Unknown",
      "question": "山田さんは心が女性ですか？",
    },
    {
      "answer": "False",
      "question": "山田さんは子供であり、保護者と一緒にいますか？",
    },
    {
      "answer": "True",
      "question": "山田さんは隠れるために女性用トイレに入りましたか？",
    },
    {
      "answer": "True",
      "question": "トイレは公園にありますか？",
    },
    {
      "answer": "False",
      "question": "トイレは山田さんの家にありますか？",
    },
    {
      "answer": "Unknown",
      "customMessage": "人目には付きづらかったと思います。",
      "question": "トイレは人目につく場所にありますか？",
    },
    {
      "answer": "False",
      "question": "女子トイレには山田さん以外に誰かいますか？",
    },
  ],
  "quiz": "山田さんは男性なのに女性用トイレにいる。それはなぜか？ （ヒントはありますか？と質問するとヒントが出ます。）
",
  "simpleTruth": "山田さんは男性用トイレにいたが、他の男性に見つかりそうになり女性用トイレに逃げ込んだ。",
  "title": "女性用トイレの男性",
  "truth": "山田さんがいるのは人気のない深夜の公園の公衆便所。 その日、山田さん人を殺害し男性用トイレに死体を隠そうとした。 しかし、その途中で山田さんはトイレに向かってくる複数の男性の喋り声に気づく。 山田さんは返り血を浴びており、もしも彼らに見つかれば間違いなく不審に思われるだろう。公園は見通しが良いためトイレから出ることも難しい。幸い、男性の位置からはトイレの入り口は見えない。 山田さんは彼らに見つからないように、男性達が入らないであろう女性用トイレに逃げ込んだ。
",
}
`;

exports[`getStory > publicOrWithAuthor > should return the story if it is authored by the author 1`] = `
{
  "author": {
    "id": "should return the story if it is authored by the author-authorId",
    "name": null,
  },
  "id": "should return the story if it is authored by the author-storyId",
  "published": false,
  "publishedAt": 1724853281065,
  "questionExamples": [
    {
      "answer": "True",
      "customMessage": "いい質問ですね！山田さんは人を殺しています。",
      "question": "山田さんは人を殺害しましたか？",
    },
    {
      "answer": "Unknown",
      "question": "山田さんは心が女性ですか？",
    },
    {
      "answer": "False",
      "question": "山田さんは子供であり、保護者と一緒にいますか？",
    },
    {
      "answer": "True",
      "question": "山田さんは隠れるために女性用トイレに入りましたか？",
    },
    {
      "answer": "True",
      "question": "トイレは公園にありますか？",
    },
    {
      "answer": "False",
      "question": "トイレは山田さんの家にありますか？",
    },
    {
      "answer": "Unknown",
      "customMessage": "人目には付きづらかったと思います。",
      "question": "トイレは人目につく場所にありますか？",
    },
    {
      "answer": "False",
      "question": "女子トイレには山田さん以外に誰かいますか？",
    },
  ],
  "quiz": "山田さんは男性なのに女性用トイレにいる。それはなぜか？ （ヒントはありますか？と質問するとヒントが出ます。）
",
  "simpleTruth": "山田さんは男性用トイレにいたが、他の男性に見つかりそうになり女性用トイレに逃げ込んだ。",
  "title": "女性用トイレの男性",
  "truth": "山田さんがいるのは人気のない深夜の公園の公衆便所。 その日、山田さん人を殺害し男性用トイレに死体を隠そうとした。 しかし、その途中で山田さんはトイレに向かってくる複数の男性の喋り声に気づく。 山田さんは返り血を浴びており、もしも彼らに見つかれば間違いなく不審に思われるだろう。公園は見通しが良いためトイレから出ることも難しい。幸い、男性の位置からはトイレの入り口は見えない。 山田さんは彼らに見つからないように、男性達が入らないであろう女性用トイレに逃げ込んだ。
",
}
`;

exports[`getStory > publicOrWithAuthor > should return the story if it is public 1`] = `
{
  "author": {
    "id": "should return the story if it is public-authorId",
    "name": null,
  },
  "id": "should return the story if it is public-storyId",
  "published": true,
  "publishedAt": 1724853281065,
  "questionExamples": [
    {
      "answer": "True",
      "customMessage": "いい質問ですね！山田さんは人を殺しています。",
      "question": "山田さんは人を殺害しましたか？",
    },
    {
      "answer": "Unknown",
      "question": "山田さんは心が女性ですか？",
    },
    {
      "answer": "False",
      "question": "山田さんは子供であり、保護者と一緒にいますか？",
    },
    {
      "answer": "True",
      "question": "山田さんは隠れるために女性用トイレに入りましたか？",
    },
    {
      "answer": "True",
      "question": "トイレは公園にありますか？",
    },
    {
      "answer": "False",
      "question": "トイレは山田さんの家にありますか？",
    },
    {
      "answer": "Unknown",
      "customMessage": "人目には付きづらかったと思います。",
      "question": "トイレは人目につく場所にありますか？",
    },
    {
      "answer": "False",
      "question": "女子トイレには山田さん以外に誰かいますか？",
    },
  ],
  "quiz": "山田さんは男性なのに女性用トイレにいる。それはなぜか？ （ヒントはありますか？と質問するとヒントが出ます。）
",
  "simpleTruth": "山田さんは男性用トイレにいたが、他の男性に見つかりそうになり女性用トイレに逃げ込んだ。",
  "title": "女性用トイレの男性",
  "truth": "山田さんがいるのは人気のない深夜の公園の公衆便所。 その日、山田さん人を殺害し男性用トイレに死体を隠そうとした。 しかし、その途中で山田さんはトイレに向かってくる複数の男性の喋り声に気づく。 山田さんは返り血を浴びており、もしも彼らに見つかれば間違いなく不審に思われるだろう。公園は見通しが良いためトイレから出ることも難しい。幸い、男性の位置からはトイレの入り口は見えない。 山田さんは彼らに見つからないように、男性達が入らないであろう女性用トイレに逃げ込んだ。
",
}
`;

exports[`getStory > withAuthorId > should return the story if it is authored by the author 1`] = `
{
  "author": {
    "id": "should return the story if it is authored by the author-authorId",
    "name": null,
  },
  "id": "should return the story if it is authored by the author-storyId",
  "published": false,
  "publishedAt": 1724853281065,
  "questionExamples": [
    {
      "answer": "True",
      "customMessage": "いい質問ですね！山田さんは人を殺しています。",
      "question": "山田さんは人を殺害しましたか？",
    },
    {
      "answer": "Unknown",
      "question": "山田さんは心が女性ですか？",
    },
    {
      "answer": "False",
      "question": "山田さんは子供であり、保護者と一緒にいますか？",
    },
    {
      "answer": "True",
      "question": "山田さんは隠れるために女性用トイレに入りましたか？",
    },
    {
      "answer": "True",
      "question": "トイレは公園にありますか？",
    },
    {
      "answer": "False",
      "question": "トイレは山田さんの家にありますか？",
    },
    {
      "answer": "Unknown",
      "customMessage": "人目には付きづらかったと思います。",
      "question": "トイレは人目につく場所にありますか？",
    },
    {
      "answer": "False",
      "question": "女子トイレには山田さん以外に誰かいますか？",
    },
  ],
  "quiz": "山田さんは男性なのに女性用トイレにいる。それはなぜか？ （ヒントはありますか？と質問するとヒントが出ます。）
",
  "simpleTruth": "山田さんは男性用トイレにいたが、他の男性に見つかりそうになり女性用トイレに逃げ込んだ。",
  "title": "女性用トイレの男性",
  "truth": "山田さんがいるのは人気のない深夜の公園の公衆便所。 その日、山田さん人を殺害し男性用トイレに死体を隠そうとした。 しかし、その途中で山田さんはトイレに向かってくる複数の男性の喋り声に気づく。 山田さんは返り血を浴びており、もしも彼らに見つかれば間違いなく不審に思われるだろう。公園は見通しが良いためトイレから出ることも難しい。幸い、男性の位置からはトイレの入り口は見えない。 山田さんは彼らに見つからないように、男性達が入らないであろう女性用トイレに逃げ込んだ。
",
}
`;
